src/Raven.Server/Documents/ShardedHandlers/ShardedDocumentHandler.cs

```diff
@@ -1,49 +1,60 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.IO;
 using System.Linq;
 using System.Net;
-using System.Net.Http;
 using System.Text;
 using System.Threading.Tasks;
 using Microsoft.Extensions.Primitives;
 using Raven.Client;
 using Raven.Client.Documents.Changes;
 using Raven.Client.Documents.Commands;
+using Raven.Server.Documents.ShardedHandlers.ShardedCommands;
 using Raven.Server.Documents.Sharding;
 using Raven.Server.Routing;
 using Raven.Server.ServerWide.Context;
 using Raven.Server.TrafficWatch;
 using Raven.Server.Utils;
 using Sparrow.Json;
-using Sparrow.Json.Parsing;
 
-namespace Raven.Server.Documents.Handlers
+namespace Raven.Server.Documents.ShardedHandlers
 {
     public class ShardedDocumentHandler : ShardedRequestHandler
     {
         [RavenShardedAction("/databases/*/docs", "HEAD")]
         public async Task Head()
         {
             var id = GetQueryStringValueAndAssertIfSingleAndNotEmpty("id");
-            var changeVector = GetStringFromHeaders("If-None-Match");
 
             using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
             {
                 var index = ShardedContext.GetShardIndex(context, id);
 
-                var cmd = new ShardedHeadCommand
-                {
-                    Id = id,
-                    ChangeVector = changeVector
-                    
-                };
+                var cmd = new ShardedHeadCommand(this, Headers.IfNonMatch);
 
                 await ShardedContext.RequestExecutors[index].ExecuteAsync(cmd, context);
                 HttpContext.Response.StatusCode = (int)cmd.StatusCode;
                 HttpContext.Response.Headers[Constants.Headers.Etag] = cmd.Result;
             }
         }
 
+        [RavenShardedAction("/databases/*/docs/size", "GET")]
+        public async Task GetDocSize()
+        {
+            var id = GetQueryStringValueAndAssertIfSingleAndNotEmpty("id");
+
+            using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var index = ShardedContext.GetShardIndex(context, id);
+
+                var cmd = new ShardedCommand(this, Headers.None);
+                await ShardedContext.RequestExecutors[index].ExecuteAsync(cmd, context);
```

Comment:
I wonder if we need to worry about race conditions with the number of shard changing midway?

Reply:
I think, this should be taken care as part of the re-sharding in the replication level? But it is a good point that we need to keep in mind

Reply:
The problem is that we have to take that into account right now
Otherwise, it will mean another set of bugs

Basically, we need to be able to get the request execs and the index as an atomic operation.

Reply:
No, that is something that we'll be handled just fine, I think. We will need two state approach here.
But the process of updates is the hard part, so we need to make sure that this works.

Having to be prepare for a shift each time is too complex and hard.